pkg/kubectl/cmd/delete.go

```diff
@@ -308,20 +329,68 @@ func DeleteResult(r *resource.Result, out io.Writer, ignoreNotFound bool, shortO
 	return nil
 }
 
-func cascadingDeleteResource(info *resource.Info, out io.Writer, shortOutput bool, mapper meta.RESTMapper) error {
+func cascadingDeleteResource(info *resource.Info, out io.Writer, shortOutput bool, mapper meta.RESTMapper, waitForDeletion bool, timeout time.Duration) error {
 	falseVar := false
 	deleteOptions := &metav1.DeleteOptions{OrphanDependents: &falseVar}
-	return deleteResource(info, out, shortOutput, mapper, deleteOptions)
-}
-
-func deleteResource(info *resource.Info, out io.Writer, shortOutput bool, mapper meta.RESTMapper, deleteOptions *metav1.DeleteOptions) error {
-	if err := resource.NewHelper(info.Client, info.Mapping).DeleteWithOptions(info.Namespace, info.Name, deleteOptions); err != nil {
-		return cmdutil.AddSourceToErr("deleting", info.Source, err)
+	obj, err := deleteResource(info, out, shortOutput, mapper, deleteOptions, !waitForDeletion)
+	if err != nil {
+		return err
+	}
+	if !waitForDeletion {
+		return nil
+	}
+	// Keep deleting the resource with uid precondition until we get a
+	// resource not found error.
+	// We use DELETE here instead of using GET to ensure that DELETE
+	// works without requiring GET permissions.
+	uid, err := getUID(obj)
+	if err != nil {
+		return fmt.Errorf("unexpected error in extracting uid: %s")
```

Comment:
> Wait by sending DELETE's without precondition.
> 
> Doing the second seems fine to me. There might be scenarios where we will end up deleting the new resource as well, but I expect them to be very very rare.  I expect users to wait for delete to return and then only create new resources.

I don't think that's a good idea. Controller-created resources in response to a deletion are common, and one client delete call shouldn't trigger two deletes. If we don't have uid info to make the delete call idempotent, I wouldn't expect to submit multiple delete calls.